src/core/net/dns_client.cpp

```diff
@@ -815,20 +862,77 @@ void Client::SendQuery(Query &aQuery, QueryInfo &aInfo, bool aUpdateTimer)
         SuccessOrExit(error = message->Append(Question(kQuestionRecordTypes[aInfo.mQueryType][num])));
     }
 
-    messageInfo.SetPeerAddr(aInfo.mConfig.GetServerSockAddr().GetAddress());
-    messageInfo.SetPeerPort(aInfo.mConfig.GetServerSockAddr().GetPort());
+    length = message->GetLength() - message->GetOffset();
+#if OPENTHREAD_CONFIG_DNS_CLIENT_OVER_TCP_ENABLE
 
-    SuccessOrExit(error = mSocket.SendTo(*message, messageInfo));
+    // Send udp query in case of message being greater than kUdpQueryMaxSize or as last resort.
+    if (aInfo.mConfig.GetTransportProto() == QueryConfig::kDnsTcpTransport)
+    {
+        mSendLink.mNext = nullptr;
+        mSendLink.mData = mSendBufferBytes;
 
-exit:
-    FreeMessageOnError(message, error);
+        // Prepending the DNS query with length of the packet according to RFC1035.
+        WriteUint16(length, mSendBufferBytes);
+        message->Read(message->GetOffset(), (mSendBufferBytes + sizeof(uint16_t)), length);
+        mSendLink.mLength = length + sizeof(uint16_t);
 
-    UpdateQuery(aQuery, aInfo);
+        mTcpTimer.Stop();
+        switch (mTcpState)
+        {
+        case kTcpUninitialized:
+            mCurrentTcpQuery = &aQuery;
+            SuccessOrExit(InitTcpSocket());
+            SuccessOrExit(mEndpoint.Connect(AsCoreType(&aInfo.mConfig.mServerSockAddr), OT_TCP_CONNECT_NO_FAST_OPEN));
+            break;
+        case kTcpInitialized:
+            if (memcmp(&mEndpoint.GetPeerAddress(), &aInfo.mConfig.mServerSockAddr, sizeof(otSockAddr)) != 0)
+            {
+                error = kErrorFailed;
+                ExitNow();
+            }
+            mCurrentTcpQuery = &aQuery;
+            mTcpState        = kTcpSendingInProgress;
+            SuccessOrExit(error = mEndpoint.SendByReference(mSendLink, /* aFlags */ 0));
+            break;
+        case kTcpConecting:
+        case kTcpSendingInProgress:
+            if (memcmp(&mEndpoint.GetPeerAddress(), &aInfo.mConfig.mServerSockAddr, sizeof(otSockAddr)) != 0)
+            {
+                error = kErrorFailed;
+                ExitNow();
+            }
+            // Decrease number of attempts on TCP query since this send will be handled in send done.
+            aInfo.mTransmissionCount--;
```

Comment:
I think the idea is to use `aInfo.mTransmissionCount` as a way to order the queries that use TCP?
I feel this is complex to follow and may be better to keep it simple.
May be simpler to use `SendByExtension` (added it in new comment).

Reply:
fixed.